### ISSUE TITLE
gsl::copy => compiler error C4996

### ISSUE THREAD

**Issue by Ziriax** [Microsoft/GSL#621]
The following code 

```
	float s[] = { 1,2,3 };
	float t[] = { 4,5,6 };
	gsl::copy(gsl::make_span(s, 3), gsl::make_span(t, 3));
```

gives the compiler error (Visual Studio 2017 15.5.6 toolset v141, GSL code from 14-feb-2018, SHA c87c123d1b3e64ae2cf725584f0c004da4d90f1c)

```
Error C4996 'std::copy_n::_Unchecked_iterators::_Deprecate': Call to 'std::copy_n' with parameters that may be unsafe - this call relies on the caller to check that the passed values are correct. To disable this warning, use -D_SCL_SECURE_NO_WARNINGS. 
```

The above is just dummy code to reproduce my problem.

What is the correct way to call `gsl::copy`?

If I look at the implementation, it already performs the following check:

```
    Expects(dest.size() >= src.size());
```

So why do I get error C4996?

Thanks a lot,
Peter Verswyvelen

**Comment by beinhaerter** [comment]
For more information you can read https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4996#unsafe-standard-library-functions.
STL does not know that GSL is doing it right, so it emits the warning.
An easy solution for you is already written in the warning: "To disable this warning, use -D_SCL_SECURE_NO_WARNINGS."

Other solutions would require a change in the VS STL or in GSL.

Regards,
Werner

**Comment by Ziriax** [comment]
Thanks for the quick reply.

Yes, I'm aware that defining the preprocessor symbol `_SCL_SECURE_NO_WARNING` suppresses these warnings, but I like have these safety checks in debug builds. 

I was assuming that `gsl::copy` was specifically made for handling this. After all, I first called `std::copy` and obviously also got the error. 

Shouldn't the  `gsl::copy` implementation use `stdext::make_checked_array_iterator` when it is compiled using the MSVC compiler? That would fix it no?

**Comment by beinhaerter** [comment]
Good idea. The following patch fixes it for me. Maybe some MS official can comment if this is the right way to do it and if I should create a pull request.

diff --git a/include/gsl/gsl_algorithm b/include/gsl/gsl_algorithm
index 710792f..b7c704b 100644
--- a/include/gsl/gsl_algorithm
+++ b/include/gsl/gsl_algorithm
@@ -51,7 +51,11 @@ void copy(span<SrcElementType, SrcExtent> src, span<DestElementType, DestExtent>
                   "Source range is longer than target range");

     Expects(dest.size() >= src.size());
+#ifdef _MSC_VER
+    std::copy_n(src.data(), src.size(), stdext::make_checked_array_iterator(dest.data(), dest.size()));
+#else
     std::copy_n(src.data(), src.size(), dest.data());
+#endif
 }

 } // namespace gsl

**Comment by neilmacintosh** [comment]
Fixed by #633 with a single line! Thanks everyone for thinking about how to do this in a minimally invasive fashion.
